### PR TITLE
STM32L4 PDMIn support

### PR DESCRIPTION
**Pull request by m-mcgowan**
Implements PDMIn for STM32L4/Swan R5 using the SAI peripheral and decimation/filtering in software.

In addition to the the PDMIn module, this PR also enables native compilation and displayio on Swan R5, which has worked fine in our testing.

### CODE REVIEW

**Comment by m-mcgowan** [comment]
Two checks in the `build-arm` job failed retrieving external dependencies due to a limit on the number of times a URL can be requested. I don't have access to re-run these checks - grateful if someone with access can re-run these failed checks.

File: ports/stm/common-hal/audiobusio/PDMIn.c [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        mp_raise_ValueError(translate("only mono is supported"));
+    }
+    if (sample_rate != 16000) {
+        mp_raise_ValueError(translate("only sample_rate=16000 is supported"));
+    }
+    if (bit_depth != 16) {
+        mp_raise_ValueError(translate("only bit_depth=16 is supported"));
+    }
+    if (oversample != 64) {
+        mp_raise_ValueError(translate("only oversample=64 is supported"));
```
Comment (dhalbert) [inline ports/stm/common-hal/audiobusio/PDMIn.c:67]: Use the `py/argcheck.c` `mp_arg_validate...()` routines here instead, if possible. This allows reuse of existing error message, saving space, and saving work for the translators. Look at some other uses of `mp_arg_validate*` in `ports/*/common-hal/`.

If there is no existing routine you can use, at least try to reuse an existing error message. Use the `%q` (for QSTR) designator in the error message, so you can pass in an `MP_QSTR_...` arg name.

File: ports/stm/boards/swan_r5/mpconfigboard.mk [hunk 2]
```diff
@@ -71,5 +70,8 @@ CIRCUITPY_BUSDEVICE = 0
 CIRCUITPY_KEYPAD = 1
 CIRCUITPY_RGBMATRIX = 0
 CIRCUITPY_RTC = 1
-
 CIRCUITPY_BUILD_EXTENSIONS = bin,uf2
+CIRCUITPY_ENABLE_MPY_NATIVE = 1
+CIRCUITPY_AUDIOBUSIO = 1
+CIRCUITPY_AUDIOBUSIO_I2SOUT = 0
+CIRCUITPY_AUDIOBUSIO_PDMIN = 1
```
Comment (dhalbert) [inline ports/stm/boards/swan_r5/mpconfigboard.mk:77]: Move these up into lines 41-46, with the other enabling/disabling values, and consider alphabetizing those lines.

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thanks for this work. I have some minor comments.

Consider redoing the copyright lines on files you have modified extensively.

File: ports/stm/common-hal/audiobusio/PDMIn.c (hunk 1 above)
Comment (m-mcgowan) [reply ports/stm/common-hal/audiobusio/PDMIn.c:67]: I can look into the changes you suggested, thanks for the pointer.  I'd much prefer to use an established pattern than something ad hoc.

Just for some background, most of the error messages were present in the nrf port, so the translations appear to be reused - the translation file now has 2 file references for the same translation text that already existed. This should mean reduced work for the translators. I can look into quick wins to reduce this further.

Comment (dhalbert) [reply ports/stm/common-hal/audiobusio/PDMIn.c:67]: If that's the case, we can leave it alone for now, and then shrink it later. I was wondering if the STM32L4 had limited flash, as well.

Comment (dhalbert) [reply ports/stm/common-hal/audiobusio/PDMIn.c:67]: So don't bother changing these now, unless you need the space.

Comment (m-mcgowan) [reply ports/stm/common-hal/audiobusio/PDMIn.c:67]: That's good to hear. Thanks!

File: ports/stm/boards/swan_r5/mpconfigboard.mk (hunk 2 above)
Comment (m-mcgowan) [reply ports/stm/boards/swan_r5/mpconfigboard.mk]: This is done - all module names in alphabetical order.

File: ports/stm/common-hal/audiobusio/PDMIn.c (hunk 1 above)
Comment (m-mcgowan) [reply ports/stm/common-hal/audiobusio/PDMIn.c:67]: Just to follow up on your question about flash space - the L4+ used in the Swan has 2MB flash, partitioned as

* 64kb - UF2 bootloader
* 960kb - runtime (presently 520kB used for a release build.)
* 1024kb - python scripts

**Comment by m-mcgowan** [comment]
@dhalbert - I've made the changes suggested. It would be great if you could approve and merge this PR.

**Review by dhalbert — APPROVED** [review]
OK, this looks good! Sorry for the delay.

**Comment by m-mcgowan** [comment]
Oh no worries, thank you for reviewing and approving.

**Pull request merged**